### PR TITLE
Set Default File Extension for `NIOLeafFiles`

### PR DESCRIPTION
**Pull request by tdotclare**
This release adds an initialization parameter to `NIOLeafFiles` to allow setting the default extension to be used for Leaf files.

```swift
public struct NIOLeafFiles: LeafSource {
    ...
    /// Initialize `NIOLeafFiles` with a NIO file IO object, limit options, and sandbox/view dirs
    /// - Parameters:
    ///   - fileio: `NonBlockingFileIO` file object
    ///   - limits: Options for constraining which files may be read - see `NIOLeafFiles.Limit`
    ///   - sandboxDirectory: Full path of the lowest directory which may be escaped to
    ///   - viewDirectory: Full path of the default directory templates are relative to
    ///   - defaultExtension: The default extension inferred files will have (defaults to `leaf`)
    ///
    /// `viewDirectory` must be contained within (or overlap) `sandboxDirectory`
    public init(fileio: NonBlockingFileIO,
                limits: Limit = .default,
                sandboxDirectory: String = "/",
                viewDirectory: String = "/",
                defaultExtension: String = "leaf") {...}
```

Usage:
```swift
    // Use "leaf4" instead of "leaf" as the implied file extension
    let source: LeafSource = NIOLeafFiles(..., defaultExtension: "leaf4")
```

### Internal Changes
* Various documentation of public protocols/objects
* General rearrangement of code for clarity

### CODE REVIEW

File: Sources/LeafKit/Data/TemplateDataEncoder.swift [hunk 1]
```diff
@@ -1,232 +0,0 @@
-// TODO: - Should this be reimplemented? This let Leaf3 use Futures as data types?
-
```
Comment (tdotclare) [inline Sources/LeafKit/Data/TemplateDataEncoder.swift:2]: Vapor/Leaf/LeafEncoder should probably be moved to LeafKit and would obviate reimplementing this file

File: Sources/LeafKit/LeafSerialize/LeafContext.swift [hunk 2]
```diff
@@ -1,3 +1,6 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
+
```
Comment (tdotclare) [inline Sources/LeafKit/LeafSerialize/LeafContext.swift:3]: Annotation Only

File: Sources/LeafKit/Deprecated.swift [hunk 3]
```diff
@@ -1,3 +1,6 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/Deprecated.swift:2]: Annotation Only

File: Sources/LeafKit/Exports.swift [hunk 4]
```diff
@@ -1,9 +1,13 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/Exports.swift:2]: Annotation Only

File: Sources/LeafKit/LeafAST.swift [hunk 5]
```diff
@@ -1,19 +1,17 @@
...
 
+    // MARK: - Internal/Private Only
     let name: String
 
-    private var rawAST: [Syntax]?
-
-    private(set) var ast: [Syntax]
-    private(set) var externalRefs = Set<String>()
-    private(set) var unresolvedRefs = Set<String>()
-    private(set) var flat: Bool
```
Comment (tdotclare) [inline Sources/LeafKit/LeafAST.swift:15]: Moved below and changed to `internal private(set)`

File: Sources/LeafKit/LeafCache/DefaultLeafCache.swift [hunk 6]
```diff
@@ -1,23 +1,28 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafCache/DefaultLeafCache.swift:2]: Annotation/Rearrangement Only

File: Sources/LeafKit/LeafCache/LeafCache.swift [hunk 7]
```diff
@@ -1,31 +1,49 @@
-// MARK: LeafCache Protocol Definition
-/// Public protocol to adhere to for storing and accessing `LeafAST`s for `LeafRenderer`
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafCache/LeafCache.swift:2]: * Annotation/Rearrangement
* `entryCount() -> Int` changed to `count: Int`

File: Sources/LeafKit/LeafData/LeafData.swift [hunk 8]
```diff
@@ -1,22 +1,16 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafData/LeafData.swift:2]: Annotation/Rearrangement Only

File: Sources/LeafKit/LeafData/LeafDataRepresentable.swift [hunk 9]
```diff
@@ -1,3 +1,6 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafData/LeafDataRepresentable.swift:2]: Annotation Only

File: Sources/LeafKit/LeafData/LeafDataStorage.swift [hunk 10]
```diff
@@ -1,3 +1,6 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafData/LeafDataStorage.swift:2]: Annotation Only

File: Sources/LeafKit/LeafLexer/LeafToken.swift [hunk 11]
```diff
@@ -75,356 +78,3 @@ public enum LeafToken: CustomStringConvertible, Equatable  {
         }
     }
 }
-
-// MARK: - `Parameter` Token Type
-
-/// An associated value enum holding data, objects or values usable as parameters to a `.tag`
```
Comment (tdotclare) [inline Sources/LeafKit/LeafLexer/LeafToken.swift:81]: Moved to LeafLexer/LeafParameterTypes.swift

File: Sources/LeafKit/LeafLexer/LeafToken.swift [hunk 12]
```diff
@@ -24,7 +27,7 @@
 /// # TODO
 /// - LeafTokens would ideally also store the range of their location in the original source template
 
-public enum LeafToken: CustomStringConvertible, Equatable  {
+internal enum LeafToken: CustomStringConvertible, Equatable  {
```
Comment (tdotclare) [inline Sources/LeafKit/LeafLexer/LeafToken.swift:30]: LeafToken made internal

File: Sources/LeafKit/LeafLexer/LeafToken.swift [hunk 13]
```diff
@@ -75,356 +78,3 @@ public enum LeafToken: CustomStringConvertible, Equatable  {
...
-        switch self {
-            case .int(let i):    return i.description
-            case .double(let d): return d.description
-        }
-    }
-}
-
-// MARK:- --- THIS SECTION TO BE MOVED TO A NEW FILE ---
-/// `ParameterDeclaration is NOT used at the lexing stage and is not presentable as a LeafToken - it
-/// is built and interpreted during LeafParser.parse(). Move to a more appropriate file.
```
Comment (tdotclare) [inline Sources/LeafKit/LeafLexer/LeafToken.swift:175]: Moved to LeafParser/LeafParameter.swift

File: Sources/LeafKit/LeafLexer/LeafToken.swift [hunk 14]
```diff
@@ -75,356 +78,3 @@ public enum LeafToken: CustomStringConvertible, Equatable  {
...
-            // .tag, .variable not atomic
-            default: return nil
-        }
-        return .raw(buffer)
-    }
-}
-// MARK: --- END OF SECTION TO BE MOVED ---
-
-// MARK: - Observational Helper Extensions to `Parameter` Types
-
```
Comment (tdotclare) [inline Sources/LeafKit/LeafLexer/LeafToken.swift:373]: Moved to LeafLexer/LeafParameterTypes.swift and made internal

File: Sources/LeafKit/LeafParser/LeafParameter.swift [hunk 15]
```diff
@@ -0,0 +1,201 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafParser/LeafParameter.swift:2]: Broken out from LeafToken.swift

File: Tests/LeafKitTests/LeafTests.swift [hunk 16]
```diff
@@ -313,8 +313,6 @@ final class LeafTests: XCTestCase {
         }
     }
```
Comment (tdotclare) [inline Tests/LeafKitTests/LeafTests.swift:314]: Removed old annotations

File: Tests/LeafKitTests/LeafKitTests.swift [hunk 17]
```diff
@@ -2,10 +2,6 @@ import XCTest
 import NIOConcurrencyHelpers
 @testable import LeafKit
```
Comment (tdotclare) [inline Tests/LeafKitTests/LeafKitTests.swift:3]: Removed old annotations and refactor using `LeafCache.entryCount()` to `.count`

File: Sources/LeafKit/LeafSyntax/LeafTag.swift [hunk 18]
```diff
@@ -1,3 +1,6 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafSyntax/LeafTag.swift:2]: Annotations Only

File: Sources/LeafKit/LeafSyntax/LeafSyntax.swift [hunk 19]
```diff
@@ -1,4 +1,7 @@
-// FIXME: - `Syntax` is overly monolithic
+// MARK: Subject to change prior to 1.0.0 release
```
Comment (tdotclare) [inline Sources/LeafKit/LeafSyntax/LeafSyntax.swift:1]: Annotations make `Syntax.ConditionalSyntax.NaturalType` internal

File: Sources/LeafKit/LeafSource/LeafSources.swift [hunk 20]
```diff
@@ -1,17 +1,32 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafSource/LeafSources.swift:2]: Annotation/Rearrangment Only

File: Sources/LeafKit/LeafSource/LeafSource.swift [hunk 21]
```diff
@@ -1,15 +1,21 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafSource/LeafSource.swift:2]: Annotation Only

File: Sources/LeafKit/LeafSerialize/ParameterResolver.swift [hunk 22]
```diff
@@ -1,17 +1,16 @@
-import Foundation
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafSerialize/ParameterResolver.swift:2]: * Annotations/Rearrangment
* Move helper structs internal to `ParameterDeclaration`

File: Sources/LeafKit/LeafParser/LeafParser.swift [hunk 23]
```diff
@@ -1,124 +1,33 @@
-extension String: Error {}
-
-private struct TagDeclaration {
```
Comment (tdotclare) [inline Sources/LeafKit/LeafParser/LeafParser.swift:3]: * Annotations/Rearrangment
* Explicitly makes `LeafParser` internal
* Consolidates and moves helper structs internal to `LeafParser`

File: Sources/LeafKit/LeafRenderer.swift [hunk 24]
```diff
@@ -1,8 +1,10 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
+
```
Comment (tdotclare) [inline Sources/LeafKit/LeafRenderer.swift:3]: Annotation/Rearrangment Only

File: Sources/LeafKit/LeafSerialize/LeafSerializer.swift [hunk 25]
```diff
@@ -1,11 +1,9 @@
-struct LeafSerializer {
-    private let ast: [Syntax]
```
Comment (tdotclare) [inline Sources/LeafKit/LeafSerialize/LeafSerializer.swift:2]: * Annotation/Rearrangment
* `LeafSerializer` made explicitly internal
* Internal helper methods made private

File: Sources/LeafKit/LeafSource/NIOLeafFiles.swift [hunk 26]
```diff
@@ -5,6 +5,17 @@ import Foundation
 ///
 /// Default initializer will
 public struct NIOLeafFiles: LeafSource {
+    // MARK: - Public
+
```
Comment (tdotclare) [inline Sources/LeafKit/LeafSource/NIOLeafFiles.swift:9]: * Annotations/Documentation/Rearrangment
* Adds `defaultExtension` parameter to initializer for `NIOLeafFiles`

File: Sources/LeafKit/LeafLexer/LeafRawTemplate.swift [hunk 27]
```diff
@@ -1,13 +1,13 @@
-/// `TemplateSource` was ambiguous - renamed to `LeafRawTemplate` and make
-/// `LeafFiles` (to also be renamed to `LeafSource`)  return it directly rather than a ByteBuffer
```
Comment (tdotclare) [inline Sources/LeafKit/LeafLexer/LeafRawTemplate.swift:2]: * Annotation/Rearrangment
* Makes `LeafRawTemplate` explicitly internal

File: Sources/LeafKit/LeafLexer/LeafParameterTypes.swift [hunk 28]
```diff
@@ -0,0 +1,142 @@
+// MARK: Subject to change prior to 1.0.0 release
+
```
Comment (tdotclare) [inline Sources/LeafKit/LeafLexer/LeafParameterTypes.swift:2]: Broken out from LeafTemplate.swift

File: Sources/LeafKit/LeafLexer/LeafLexer.swift [hunk 29]
```diff
@@ -1,10 +1,36 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafLexer/LeafLexer.swift:2]: * Annotation/Rearrangment
* Makes `LeafLexer` implicitly internal

File: Sources/LeafKit/LeafError.swift [hunk 30]
```diff
@@ -1,3 +1,6 @@
+// MARK: Subject to change prior to 1.0.0 release
+// MARK: -
```
Comment (tdotclare) [inline Sources/LeafKit/LeafError.swift:2]: * Annotation/Rearrangment
* Makes `LexerError.lexed` internal to allow `LeafToken` to be internal
* Renames `LexerError.internalError` to `LexerError.unknownError`

**Comment by tanner0101** [comment]
These changes are now available in [1.0.0-rc.1.12](https://github.com/vapor/leaf-kit/releases/tag/1.0.0-rc.1.12)

**Pull request merged**